[llm] Add a generic text only LLM runner

larryliu0820 opened the pull request:
Introducing `text_llm_runner`. This can be used to run all text only decoder only LLM models supported by ExecuTorch.

* Metadata is being read out from the .pte file and being used to construct the runner object.
* examples/models/llama/runner.h[.cpp] only contains a simple wrapper around `text_llm_runner.h[.cpp]`.


In next PRs I will move examples/models/phi-3-mini/runner to use the generic runner.

Will look into QNN and MediaTek runners as well.

Differential Revision: [D75910889](https://our.internmc.facebook.com/intern/diff/D75910889/)

ghstack-source-id: 288004703
Pull Request resolved: https://github.com/pytorch/executorch/pull/11342

### Summary
[PLEASE REMOVE] See [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests) for ExecuTorch PR guidelines.

[PLEASE REMOVE] If this PR closes an issue, please add a `Fixes #<issue-id>` line.

[PLEASE REMOVE] If this PR introduces a fix or feature that should be the upcoming release notes, please add a "Release notes: <area>" label. For a list of available release notes labels, check out [CONTRIBUTING.md's Pull Requests](https://github.com/pytorch/executorch/blob/main/CONTRIBUTING.md#pull-requests).

### Test plan
[PLEASE REMOVE] How did you test this PR? Please write down any manual commands you used and note down tests that you have written if applicable.

mergennachin commented on extension/llm/runner/text_llm_runner.h:
```diff
@@ -0,0 +1,109 @@
+    size_t bos_token_index = 0,
+    size_t eos_token_index = 1);
+
+std::unordered_map<std::string, int64_t> get_llm_metadata(
```
same...

mergennachin commented on extension/llm/runner/text_llm_runner.h:
```diff
@@ -0,0 +1,109 @@
+  float temperature_ = -1.0f;
+};
+
+std::unique_ptr<tokenizers::Tokenizer> load_tokenizer(
+    const std::string& tokenizer_path,
+    std::unique_ptr<std::vector<std::string>> special_tokens = nullptr,
+    std::optional<std::string> pattern = std::nullopt,
+    size_t bos_token_index = 0,
+    size_t eos_token_index = 1);
+
+std::unordered_map<std::string, int64_t> get_llm_metadata(
+    tokenizers::Tokenizer* tokenizer,
+    Module* module);
+
+std::unordered_set<uint64_t> get_eos_ids(
+    tokenizers::Tokenizer* tokenizer,
+    Module* module);
+
+std::unique_ptr<TextLLMRunner> create_text_llm_runner(
+    const std::string& model_path,
+    std::unique_ptr<::tokenizers::Tokenizer> tokenizer,
+    std::optional<const std::string> data_path = std::nullopt,
+    float temperature = -1.0f);
+
```
if you want to expose, please tag as ET_EXPERIMENTAL.

the current ET_EXPERIMENTAL is only applicable to the class

mergennachin reviewed (commented):
See inline

larryliu0820 commented on extension/llm/runner/text_llm_runner.h:
```diff
@@ -0,0 +1,109 @@
+  void stop() override;
+
+ private:
+  bool shouldStop_{false};
```
Yes android and ios are using the `stop()` API

mergennachin commented on extension/llm/runner/text_llm_runner.h:
```diff
@@ -0,0 +1,183 @@
+ */
+ET_EXPERIMENTAL std::unique_ptr<tokenizers::Tokenizer> load_tokenizer(
+    const std::string& tokenizer_path,
+    std::unique_ptr<std::vector<std::string>> special_tokens = nullptr,
```
instead of using smart pointer, how about this? 

```
const std::optional<std::vector<std::string>>& tokens
```

mergennachin replied:
That way, you're passing by const reference and not using any pointers

mergennachin commented on extension/llm/runner/text_llm_runner.h:
```diff
@@ -0,0 +1,183 @@
+      std::unique_ptr<TextPrefiller> text_prefiller,
+      std::unique_ptr<TextTokenGenerator> text_token_generator,
+      std::unique_ptr<Stats> stats,
+      float temperature = -1.0f);
```
if it's deprecated, why include this as part of the API?

If there's a user code, they can always just use create_text_llm_runner, or can we migrate the user code to use GenerationConfig directly?

mergennachin commented on examples/models/llama/runner/runner.h:
```diff
@@ -19,74 +19,20 @@
-          text_token_generator,
-      std::unique_ptr<::executorch::extension::llm::Stats> stats,
-      float temperature = -1.0f);
+std::unique_ptr<llm::TextLLMRunner> create_llama_runner(
```
do you think we should keep example::create_llama_runner method at all? 

should we just replace with llm::load_tokenizer and llm::create_llm_runner  for all the current callsites of example::create_llama_runner

It seems unnecessary indirection, and we would to showcase these low-level APIs in our examples, as opposed to having, yet another indirection (i.e., example::create_llama_runner)

i'm fine if you want to this in a follow-up PR, but I think we should do it very soon.

larryliu0820 replied:
Yeah don't have a strong opinion on this. I can do a follow up.

mergennachin reviewed (approved):
See inline @larryliu0820 and thank you!

larryliu0820 committed:
[llm] Add a generic text only LLM runner

Pull Request resolved: https://github.com/pytorch/executorch/pull/11342

Introducing `text_llm_runner`. This can be used to run all text only decoder only LLM models supported by ExecuTorch.

* Metadata is being read out from the .pte file and being used to construct the runner object.
* examples/models/llama/runner.h[.cpp] only contains a simple wrapper around `text_llm_runner.h[.cpp]`.


In next PRs I will move examples/models/phi-3-mini/runner to use the generic runner.

Will look into QNN and MediaTek runners as well.
ghstack-source-id: 288571542
@exported-using-ghexport

Differential Revision: [D75910889](https://our.internmc.facebook.com/intern/diff/D75910889/)

larryliu0820 commented:
Close, duplicate of #11342
